File: test_sender.py
#!/usr/bin/env python3

import argparse
import socket
import struct
import time
import math


def generate_continuous_sine_wave(frequency: float, sample_rate: int, duration: float, amplitude: float = 0.3):
    """Generate continuous sine wave audio samples"""
    num_samples = int(sample_rate * duration)
    samples = []
    
    for i in range(num_samples):
        t = i / sample_rate
        sample = amplitude * math.sin(2 * math.pi * frequency * t)
        # Convert to 16-bit signed integer
        sample_int = int(sample * 32767)
        # Clamp to 16-bit range
        sample_int = max(-32768, min(32767, sample_int))
        samples.append(sample_int)
    
    return samples


def send_audio_packets(host: str, port: int, sample_rate: int = 16000, 
                      tone_frequency: float = 440.0, packet_duration: float = 0.02,
                      buffer_duration: float = 3.0):
    """Send UDP audio packets with a continuous sine wave tone"""
    
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    
    print(f"Sending audio packets to {host}:{port}")
    print(f"Sample rate: {sample_rate} Hz")
    print(f"Tone frequency: {tone_frequency} Hz")
    print(f"Packet duration: {packet_duration} seconds")
    print(f"Buffer duration: {buffer_duration} seconds")
    print(f"Frame format: [2-byte seq#][4-byte sample timestamp][audio samples]")
    print("Press Ctrl+C to stop")
    
    # Calculate packet parameters
    samples_per_packet = int(sample_rate * packet_duration)
    buffer_samples = int(sample_rate * buffer_duration)
    packets_per_buffer = buffer_samples // samples_per_packet
    
    print(f"Samples per packet: {samples_per_packet}")
    print(f"Packets per buffer: {packets_per_buffer}")
    
    try:
        packet_count = 0
        seq_num = 0
        sample_timestamp = 0
        buffer_start_time = 0.0  # Time offset for continuous sine wave
        
        while True:
            # Generate continuous audio buffer
            # print(f"Generating {buffer_duration}s audio buffer...")
            audio_buffer = generate_continuous_sine_wave(
                tone_frequency, sample_rate, buffer_duration
            )
            
            # Send packets from the buffer
            for packet_idx in range(packets_per_buffer):
                start_sample = packet_idx * samples_per_packet
                end_sample = start_sample + samples_per_packet
                packet_samples = audio_buffer[start_sample:end_sample]
                
                # Pack frame header: [2 bytes seq#][4 bytes sample_timestamp]
                header = struct.pack('<HI', seq_num, sample_timestamp)
                
                # Pack audio samples as little-endian 16-bit integers
                audio_data = struct.pack(f'<{len(packet_samples)}h', *packet_samples)
                
                # Combine header and audio data
                packet_data = header + audio_data
                
                # Send packet
                sock.sendto(packet_data, (host, port))
                packet_count += 1
                
                # Update counters for next packet
                seq_num = (seq_num + 1) % 65536  # Wrap at 16-bit boundary
                sample_timestamp += len(packet_samples)
                
                # if packet_count % 50 == 0:  # Print status every 50 packets
                #     print(f"Sent {packet_count} packets (seq: {seq_num-1}, timestamp: {sample_timestamp-len(packet_samples)})...")
                
                # Sleep to maintain real-time rate
            
            # Update buffer start time for next continuous buffer
            buffer_start_time += buffer_duration
            # time.sleep(packet_duration)
            
    except KeyboardInterrupt:
        print(f"\nSent {packet_count} total packets")
        print(f"Final sequence number: {seq_num-1}")
        print(f"Final sample timestamp: {sample_timestamp}")
    finally:
        sock.close()


def main():
    parser = argparse.ArgumentParser(description="UDP Audio Packet Sender (Test Tool)")
    parser.add_argument('host', help='Target host/IP address')
    parser.add_argument('port', type=int, help='Target UDP port')
    parser.add_argument('--sample-rate', type=int, default=16000,
                       help='Audio sample rate in Hz (default: 16000)')
    parser.add_argument('--frequency', type=float, default=440.0,
                       help='Sine wave frequency in Hz (default: 440.0)')
    parser.add_argument('--packet-duration', type=float, default=0.02,
                       help='Duration of each packet in seconds (default: 0.02)')
    parser.add_argument('--buffer-duration', type=float, default=3.0,
                       help='Duration of continuous audio buffer in seconds (default: 3.0)')
    
    args = parser.parse_args()
    
    # Validate arguments
    if args.port < 1 or args.port > 65535:
        print("Error: Port must be between 1 and 65535")
        return 1
    
    if args.sample_rate <= 0:
        print("Error: Sample rate must be positive")
        return 1
    
    if args.frequency <= 0:
        print("Error: Frequency must be positive")
        return 1
    
    if args.packet_duration <= 0:
        print("Error: Packet duration must be positive")
        return 1
    
    send_audio_packets(args.host, args.port, args.sample_rate, 
                      args.frequency, args.packet_duration, args.buffer_duration)
    
    return 0


if __name__ == "__main__":
    exit(main())

File: Cpp/src/test_sender.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
#include <thread>
#include <chrono>
#include <cstring>
#include <algorithm>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#ifdef _WIN32
#include <winsock2.h>
#include <ws2tcpip.h>
#pragma comment(lib, "ws2_32.lib")
#else
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#endif

class UDPTestSender {
public:
    UDPTestSender(const std::string& host, int port, int sampleRate = 16000, 
                  double frequency = 440.0, double packetDuration = 0.02)
        : host_(host), port_(port), sampleRate_(sampleRate), 
          frequency_(frequency), packetDuration_(packetDuration) {
#ifdef _WIN32
        socket_ = INVALID_SOCKET;
#else
        socket_ = -1;
#endif
    }

    ~UDPTestSender() {
        cleanup();
    }

    bool initialize() {
#ifdef _WIN32
        // Initialize Winsock
        WSADATA wsaData;
        int result = WSAStartup(MAKEWORD(2, 2), &wsaData);
        if (result != 0) {
            std::cerr << "WSAStartup failed: " << result << std::endl;
            return false;
        }

        // Create socket
        socket_ = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
        if (socket_ == INVALID_SOCKET) {
            std::cerr << "Socket creation failed: " << WSAGetLastError() << std::endl;
            WSACleanup();
            return false;
        }
#else
        socket_ = socket(AF_INET, SOCK_DGRAM, 0);
        if (socket_ < 0) {
            perror("Socket creation failed");
            return false;
        }
#endif

        // Setup destination address
        addr_.sin_family = AF_INET;
        addr_.sin_port = htons(static_cast<uint16_t>(port_));
        
#ifdef _WIN32
        if (inet_pton(AF_INET, host_.c_str(), &addr_.sin_addr) <= 0) {
            std::cerr << "Invalid address: " << host_ << std::endl;
            return false;
        }
#else
        if (inet_pton(AF_INET, host_.c_str(), &addr_.sin_addr) <= 0) {
            perror("Invalid address");
            return false;
        }
#endif

        return true;
    }

    void sendAudioPackets() {
        if (!initialize()) {
            std::cerr << "Failed to initialize sender" << std::endl;
            return;
        }

        std::cout << "Sending audio packets to " << host_ << ":" << port_ << std::endl;
        std::cout << "Sample rate: " << sampleRate_ << " Hz" << std::endl;
        std::cout << "Tone frequency: " << frequency_ << " Hz" << std::endl;
        std::cout << "Packet duration: " << packetDuration_ << " seconds" << std::endl;
        std::cout << "Frame format: [2-byte seq#][4-byte sample timestamp][audio samples]" << std::endl;
        std::cout << "Press Ctrl+C to stop" << std::endl;

        int samplesPerPacket = static_cast<int>(sampleRate_ * packetDuration_);
        uint16_t sequenceNumber = 0;
        uint32_t sampleTimestamp = 0;
        uint64_t packetCount = 0;

        auto startTime = std::chrono::high_resolution_clock::now();
        auto nextPacketTime = startTime;

        try {
            while (true) {
                // Generate sine wave samples for this packet
                std::vector<int16_t> samples = generateSineWave(samplesPerPacket, sampleTimestamp);

                // Create packet: [2 bytes seq][4 bytes timestamp][audio samples]
                std::vector<uint8_t> packet(6 + samples.size() * 2);
                
                // Pack header (little-endian)
                std::memcpy(packet.data(), &sequenceNumber, 2);
                std::memcpy(packet.data() + 2, &sampleTimestamp, 4);
                
                // Pack audio samples (little-endian)
                std::memcpy(packet.data() + 6, samples.data(), samples.size() * 2);

                // Send packet
#ifdef _WIN32
                int result = sendto(socket_, reinterpret_cast<const char*>(packet.data()), 
                                  static_cast<int>(packet.size()), 0,
                                  reinterpret_cast<const sockaddr*>(&addr_), sizeof(addr_));
                if (result == SOCKET_ERROR) {
                    std::cerr << "Send failed: " << WSAGetLastError() << std::endl;
                    break;
                }
#else
                ssize_t result = sendto(socket_, packet.data(), packet.size(), 0,
                                      reinterpret_cast<const sockaddr*>(&addr_), sizeof(addr_));
                if (result < 0) {
                    perror("Send failed");
                    break;
                }
#endif

                // Update counters
                sequenceNumber++;
                sampleTimestamp += samplesPerPacket;
                packetCount++;

                // Status update every 50 packets
                if (packetCount % 50 == 0) {
                    std::cout << "Sent " << packetCount << " packets (seq: " 
                              << (sequenceNumber - 1) << ", timestamp: " 
                              << (sampleTimestamp - samplesPerPacket) << ")..." << std::endl;
                }

                // Wait for next packet time
                nextPacketTime += std::chrono::duration_cast<std::chrono::high_resolution_clock::duration>(
                    std::chrono::duration<double>(packetDuration_));
                
                std::this_thread::sleep_until(nextPacketTime);
            }
        } catch (const std::exception& e) {
            std::cerr << "Error: " << e.what() << std::endl;
        }

        std::cout << "\nSent " << packetCount << " total packets" << std::endl;
        std::cout << "Final sequence number: " << (sequenceNumber - 1) << std::endl;
        std::cout << "Final sample timestamp: " << (sampleTimestamp - samplesPerPacket) << std::endl;
    }

private:
    std::vector<int16_t> generateSineWave(int numSamples, uint32_t startingSampleIndex) {
        std::vector<int16_t> samples(numSamples);
        const double amplitude = 0.3;
        
        for (int i = 0; i < numSamples; ++i) {
            double t = static_cast<double>(startingSampleIndex + i) / sampleRate_;
            double sample = amplitude * std::sin(2.0 * M_PI * frequency_ * t);
            
            // Convert to 16-bit signed integer
            int32_t sampleInt = static_cast<int32_t>(sample * 32767.0);
            // Clamp to 16-bit range
            if (sampleInt > 32767) sampleInt = 32767;
            if (sampleInt < -32768) sampleInt = -32768;
            samples[i] = static_cast<int16_t>(sampleInt);
        }
        
        return samples;
    }

    void cleanup() {
#ifdef _WIN32
        if (socket_ != INVALID_SOCKET) {
            closesocket(socket_);
            socket_ = INVALID_SOCKET;
            WSACleanup();
        }
#else
        if (socket_ >= 0) {
            close(socket_);
            socket_ = -1;
        }
#endif
    }

    std::string host_;
    int port_;
    int sampleRate_;
    double frequency_;
    double packetDuration_;

#ifdef _WIN32
    SOCKET socket_;
#else
    int socket_;
#endif
    sockaddr_in addr_;
};

void printUsage(const char* programName) {
    std::cout << "Usage: " << programName << " <host> <port> [options]" << std::endl;
    std::cout << "Options:" << std::endl;
    std::cout << "  --sample-rate <rate>      Audio sample rate in Hz (default: 16000)" << std::endl;
    std::cout << "  --frequency <freq>        Sine wave frequency in Hz (default: 440.0)" << std::endl;
    std::cout << "  --packet-duration <dur>   Duration of each packet in seconds (default: 0.02)" << std::endl;
    std::cout << "  --help                   Show this help message" << std::endl;
    std::cout << std::endl;
    std::cout << "Examples:" << std::endl;
    std::cout << "  " << programName << " localhost 8000" << std::endl;
    std::cout << "  " << programName << " 192.168.1.100 8000 --frequency 880" << std::endl;
    std::cout << "  " << programName << " localhost 8000 --sample-rate 44100 --packet-duration 0.01" << std::endl;
}

int main(int argc, char* argv[]) {
    // Default parameters
    std::string host;
    int port = 0;
    int sampleRate = 16000;
    double frequency = 440.0;
    double packetDuration = 0.02;

    // Parse command line arguments
    if (argc < 3) {
        printUsage(argv[0]);
        return 1;
    }

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];

        if (arg == "--help" || arg == "-h") {
            printUsage(argv[0]);
            return 0;
        } else if (arg == "--sample-rate") {
            if (i + 1 >= argc) {
                std::cerr << "Error: --sample-rate requires a value" << std::endl;
                return 1;
            }
            try {
                sampleRate = std::stoi(argv[++i]);
                if (sampleRate <= 0) {
                    std::cerr << "Error: Sample rate must be positive" << std::endl;
                    return 1;
                }
            } catch (const std::exception& e) {
                std::cerr << "Error: Invalid sample rate: " << argv[i] << std::endl;
                return 1;
            }
        } else if (arg == "--frequency") {
            if (i + 1 >= argc) {
                std::cerr << "Error: --frequency requires a value" << std::endl;
                return 1;
            }
            try {
                frequency = std::stod(argv[++i]);
                if (frequency <= 0) {
                    std::cerr << "Error: Frequency must be positive" << std::endl;
                    return 1;
                }
            } catch (const std::exception& e) {
                std::cerr << "Error: Invalid frequency: " << argv[i] << std::endl;
                return 1;
            }
        } else if (arg == "--packet-duration") {
            if (i + 1 >= argc) {
                std::cerr << "Error: --packet-duration requires a value" << std::endl;
                return 1;
            }
            try {
                packetDuration = std::stod(argv[++i]);
                if (packetDuration <= 0) {
                    std::cerr << "Error: Packet duration must be positive" << std::endl;
                    return 1;
                }
            } catch (const std::exception& e) {
                std::cerr << "Error: Invalid packet duration: " << argv[i] << std::endl;
                return 1;
            }
        } else if (host.empty()) {
            host = arg;
        } else if (port == 0) {
            try {
                port = std::stoi(arg);
                if (port < 1 || port > 65535) {
                    std::cerr << "Error: Port must be between 1 and 65535" << std::endl;
                    return 1;
                }
            } catch (const std::exception& e) {
                std::cerr << "Error: Invalid port: " << arg << std::endl;
                return 1;
            }
        } else {
            std::cerr << "Error: Unknown argument: " << arg << std::endl;
            printUsage(argv[0]);
            return 1;
        }
    }

    if (host.empty() || port == 0) {
        std::cerr << "Error: Host and port are required" << std::endl;
        printUsage(argv[0]);
        return 1;
    }

    // Create and run sender
    UDPTestSender sender(host, port, sampleRate, frequency, packetDuration);
    sender.sendAudioPackets();

    return 0;
}